File: samples/src/commonMain/kotlin/sample/_class/sealed/SealedClass.kt
package sample._class.sealed

import deezer.kustom.KustomExportSamples

@KustomExportSamples
sealed class SealedParent(val ctorParam: Long) {
    abstract val prop: Int
    val hardcoded: Long = 42
    fun computeFoo() = 12345
}
// Should generate a class wrapper with 'protected' constructor (package visibility, broken with package erasure?)

@KustomExportSamples
class SealedChild1(override val prop: Int, ctorParam: Long) : SealedParent(ctorParam) {
    val child1Field: String = "child1"
    fun special() = "$prop-$ctorParam-$child1Field"
}

@KustomExportSamples
class SealedChild2(override var prop: Int) : SealedParent(94949) {
    val child2Field: String = "child2"
}

@KustomExportSamples
class SealedClassConsumer {
    fun consume(sealed: SealedParent): String {
        return when (sealed) {
            is SealedChild1 -> "child1:${sealed.child1Field}"
            is SealedChild2 -> "child2:${sealed.child2Field}"
        }
    }
}

File: samples/src/commonMain/kotlin/sample/_class/sealed/SealedClass.ts
import { runTest } from "../../shared_ts/RunTest"
import { assert } from "../../shared_ts/Assert"
import { assertQuiet } from "../../shared_ts/Assert"
import { sample } from '@kustom/Samples'

runTest("SealedClass", () : void => {
    var sealed1 = new sample._class.sealed.js.SealedChild1(33, 101)
    var sealed2 = new sample._class.sealed.js.SealedChild2(28)
    var extSealed = new sample._class.sealed.js.ExtendedSealedChild(28, "bliblibli")

    var consumer = new sample._class.sealed.js.SealedClassConsumer()
    assert(consumer.consume(sealed1) == "child1:child1", "Sealed class 1")
    assert(consumer.consume(sealed2) == "child2:child2", "Sealed class 2")
    assert(consumer.consume(extSealed) == "child2:child2", "Extended sealed class")

    assertQuiet(sealed1.prop == 33, "SealedChild1:field")
    assertQuiet(sealed1.ctorParam == 101, "SealedChild1:ctorParam")
    assertQuiet(sealed1.hardcoded == 42, "SealedChild1:hardcoded") // Value from Kotlin
    assertQuiet(sealed1.computeFoo() == 12345, "SealedChild1:computeFoo()") // Value from Kotlin
    assertQuiet(sealed1.special() == "33-101-child1", "SealedChild1:special()") // Compute value from both sides

    assertQuiet(sealed2.prop == 28, "SealedChild2:field")
    assertQuiet(sealed2.ctorParam == 94949, "SealedChild2:ctorParam") // Value from Kotlin
    assertQuiet(sealed2.hardcoded == 42, "SealedChild2:hardcoded") // Value from Kotlin
    assertQuiet(sealed2.computeFoo() == 12345, "SealedChild2:computeFoo()") // Value from Kotlin
})